### PR TITLE
Prohibit duplicates in gs.DETS and during event bundling

### PR DESCRIPTION
**Pull request by danielballan**
We can do this elsewhere as well (in the RunEngine, perhaps also in the scan `detectors` setters) but this is a start.

### CODE REVIEW

**Comment by tacaswell** [comment]
This does not check during bundling, only at the global state level.

**Comment by danielballan** [comment]
@tacaswell I changed the title before pushing the commits. :- ) Check again.

**Comment by danielballan** [comment]
Closes #153

**Comment by danielballan** [comment]
What is "QuantifiedCode" and why is it yelling at me?

File: bluesky/run_engine.py [hunk 1]
```diff
@@ -734,7 +734,16 @@ def _read(self, msg):
         obj = msg.obj
         self._objs_read.append(obj)
         if obj not in self._describe_cache:
-            self._describe_cache[obj] = obj.describe()
+            # Validate that there is no data key name collision.
+            data_keys = obj.describe()
+            keys = data_keys.keys()  # that is, field names
+            for known_obj, known_data_keys in self._describe_cache.items():
+                known_keys = known_data_keys.keys()  # that is, field names
+                if len(set(known_keys) & set(keys)) > 0:
```
Comment (tacaswell) [inline bluesky/run_engine.py]: Don't really need the `>0`

File: bluesky/run_engine.py [hunk 2]
```diff
@@ -734,7 +734,16 @@ def _read(self, msg):
         obj = msg.obj
         self._objs_read.append(obj)
         if obj not in self._describe_cache:
-            self._describe_cache[obj] = obj.describe()
+            # Validate that there is no data key name collision.
+            data_keys = obj.describe()
+            keys = data_keys.keys()  # that is, field names
+            for known_obj, known_data_keys in self._describe_cache.items():
```
Comment (tacaswell) [inline bluesky/run_engine.py]: This is a _bit_ more restrictive than we need it to be, we should be able to use objects with conflicting key names in the same scan as long as we don't bundle them into the same event.  On the other hand, we will deal with that complication when we find a beamline that needs it.

I like that this is a one time check.

File: bluesky/run_engine.py (hunk 1 above)
Comment (danielballan) [inline bluesky/run_engine.py]: done

File: bluesky/run_engine.py (hunk 2 above)
Comment (danielballan) [inline bluesky/run_engine.py]: I agree.

**Comment by danielballan** [comment]
@tacaswell I took

> On the other hand, we will deal with that complication when we find a beamline that needs it.

to mean, "Good enough." Satisfied with this?

**Pull request merged**